core/typeSugar.ml

```diff
@@ -4070,6 +4093,31 @@ and type_binding : context -> binding -> binding * context * usagemap =
                              (usages body)
             else () in
 
+          (* Check that quantifiers have not escaped into the typing context *)
+          let check_escaped_quantifiers quantifiers env =
+            let quantifier_set = IntSet.of_list (List.map fst quantifiers) in
+            (* Note that `type_predicate` returns true iff *all* child nodes of
+             * the type satisfy the predicate. Thus the checker returns true if
+             * *all* type variables are *not* in quantifier_set *)
+            let checker = object(_self)
+              inherit Types.type_predicate
+              method! var_satisfies (i, _, _) = not (IntSet.mem i quantifier_set)
+            end in
+            let (is_safe, _) = checker#predicates in
+            let escapees =
+              Env.fold (fun name dt acc ->
```

Comment:
Random thought: having `Env.filter` and `Env.filterMap` would be useful, though at the moment it seems that `filter` would only be used here, and `filterMap` in one place in `types.ml`